backend/api/src/http/endpoints/user.rs

```diff
@@ -1,28 +1,239 @@
 use crate::{
-    db::{
-        self,
-        user::{register, register_google_oauth},
-    },
+    db::{self, user::upsert_profile},
     error,
-    extractor::OAuthSignupToken,
-    token::{OAuthProvider, TokenSource},
+    extractor::{SessionPutProfile, TokenUserWithPurposedSession},
+    service::{mail, ServiceData},
+    token::SessionMask,
 };
-use crate::{extractor::TokenUser, token::create_signin_token};
+use crate::{extractor::TokenUser, token::create_auth_token};
+use actix_http::error::BlockingError;
 use actix_web::HttpResponse;
+use argon2::{password_hash::SaltString, Argon2, PasswordHasher};
+use chrono::{Duration, Utc};
 use core::settings::RuntimeSettings;
 use paperclip::actix::{
     api_v2_operation,
     web::{Data, Json, Query, ServiceConfig},
+    NoContent,
 };
+use rand::thread_rng;
+use sendgrid::v3::Email;
 use shared::{
     api::endpoints::{
-        user::{Profile, Register, UserLookup},
+        user::{Create, Profile, PutProfile, UserLookup, VerifyEmail},
         ApiEndpoint,
     },
-    domain::{auth::RegisterRequest, session::CreateSessionSuccess, user::UserLookupQuery},
-    error::auth::RegisterErrorKind,
+    domain::{
+        session::NewSessionResponse,
+        user::{PutProfileRequest, UserLookupQuery, VerifyEmailRequest},
+    },
 };
-use sqlx::PgPool;
+use sqlx::{Acquire, PgConnection, PgPool};
+use uuid::Uuid;
+
+async fn send_verification_email(
+    txn: &mut PgConnection,
+    user_id: Uuid,
+    email_address: String,
+    mail: &mail::Client,
+    pages_url: &str,
+) -> Result<(), error::Service> {
+    let session = db::session::create(
+        &mut *txn,
+        user_id,
+        Some(&(Utc::now() + Duration::hours(1))),
+        SessionMask::VERIFY_EMAIL,
+        None,
+    )
+    .await?;
+
+    let template = mail
+        .signup_verify_template()
+        .map_err(error::Service::DisabledService)?;
+
+    let email_link = format!("{}/verify-email/{}", pages_url, session);
+
+    mail.send_signup_verify(template, Email::new(email_address), email_link)
+        .await?;
+
+    Ok(())
+}
+
+/// Create a user
+#[api_v2_operation]
+async fn create_user(
+    config: Data<RuntimeSettings>,
+    mail: ServiceData<mail::Client>,
+    db: Data<PgPool>,
+    req: Json<<Create as ApiEndpoint>::Req>,
+) -> Result<NoContent, error::Service> {
+    let req = req.into_inner();
+
+    if req.password.is_empty() {
+        todo!("empty password is error");
+    }
+
+    let mut txn = db.begin().await?;
+
+    // todo: handle duplicate emails
+
+    let user = sqlx::query!(r#"insert into "user" default values returning id"#)
+        .fetch_one(&mut txn)
+        .await?;
+
+    let pass = req.password;
+
+    let res = actix_web::web::block(move || {
+        let password_hasher = Argon2::default();
+
+        let salt = SaltString::generate(thread_rng());
+        password_hasher
+            .hash_password(
```

Comment:
Use `hash_password_simple`

Reply:
I was recommended to put the parameters at a higher level (which is how I got the ones I'm using), as you may recall from a conversation on slack.